### DISCUSSION TITLE
Opening out-of-scope URLs in default browser closing the PWA is unexpected

### DISCUSSION THREAD

**Discussion by prettyv** [filips123/PWAsForFirefox#246] [General]
Judging by the wiki I understand that the current behaviour of closing the active PWA (or its active tab when multiple tab support is enabled) is intentional. I was surprised about this and expected it to behave just like the right-click menu option "Open Link in default browser". Is there room to adjust this behaviour or provide an additional option to not close the current tab? I'm curious about the reasons for the current default.

Also I'm unclear about what settings are actually valid or needed in combination with multiple tab support. At the moment when toggling the out-of-scope option "Force links into a new window" is selected automatically and when toggling the multiple-tabs option "Force links into a new tab" is selected. So depending on the order in which these are checked different link target behaviours are selected and it isn't immediately clear that there should be a strong coupling to either tabs or out-of-scope link handling. Unselecting either will also enable "Force links into the current tab" again regardless of whether another option is still active, adding to the potential for confusion.

What I wanted to achieve in effect was to have both multiple tabs support and opening out-of-scope links in the default browser (without closing the active tab), and otherwise forcing links into a new tab in case in-scope or explicitly allowlisted links had a `target="_blank"` attribute. Reading https://github.com/filips123/PWAsForFirefox/issues/94 it sounds like there might have been issues implementing that? Would be great if this could be made to work!

**Comment by filips123** [comment]
Ideally, "Open out-of-scope URLs in default browser" should not even open those URLs in the PWA browser, but just directly open them in a default one. So, when you clicked on some out-of-scope URL, it would directly open a new tab in your default browser.

The problem is that I couldn't find a way to intercept the link opening without creating a new tab, so I just let the new tab open, open it again in the main browser, and close the PWA tab. The reason why that tab is closed is that otherwise, you would have that website opened twice, which is not what that option intends to do.

I have some ideas how to fix this (and some other related issues), but I haven't checked yet if it works and can't promise when it will be implemented.

> Is there room to adjust this behaviour or provide an additional option to not close the current tab?

I think it's possible to make an option to prevent closing the current tab. However, this means that the out-of-scope website will be opened twice, in the PWA browser and in the main one. Is this what you intend?

> Also I'm unclear about what settings are actually valid or needed in combination with multiple tab support. 

Both "Force links into a new window" and "Force links into a new tab" should work (since 2.1.0) when tabs support and opening out-of-scope URLs are enabled together. In the first case, a new window will open and immediately close if it's not in scope, and in the second, it will be a tab. In this case, different behaviours when enabling options in a different order shouldn't really matter, because it's mostly just based on the user's preference.

> Unselecting either will also enable "Force links into the current tab" again regardless of whether another option is still active, adding to the potential for confusion.

Here, I agree that it can be confusing. I'll probably fix this so "Force links into the current tab" is only enabled if neither of the tabs support and opening out-of-scope URLs is enabled.
